vavr/src/main/java/io/vavr/concurrent/FutureImpl.java

```diff
@@ -104,50 +103,67 @@ private FutureImpl(ExecutorService executorService, Option<Try<T>> value, Queue<
      *
-     * @param executorService  An {@link ExecutorService} to run and control the computation and to perform the actions.
+     * @param executor  An {@link Executor} to run and control the computation and to perform the actions.
      * @param computation A non-blocking computation
      * @param <T> value type of the Future
      * @return a new {@code FutureImpl} instance
      */
-    static <T> FutureImpl<T> sync(ExecutorService executorService, CheckedConsumer<Predicate<Try<? extends T>>> computation) {
-        return new FutureImpl<>(executorService, Option.none(), Queue.empty(), Queue.empty(), future -> {
-            computation.accept(future::tryComplete);
-            return null;
+    static <T> FutureImpl<T> sync(Executor executor, CheckedConsumer<Predicate<Try<? extends T>>> computation) {
+        return new FutureImpl<>(executor, Option.none(), Queue.empty(), Queue.empty(), tryComplete -> {
+            computation.accept(tryComplete);
+            return null; // no thread is created in the synchronous case
         });
     }
 
     /**
      * Creates a {@code FutureImpl} that is eventually completed.
      * The given {@code computation} is <em>asynchronously</em> executed, a new thread is started.
      *
-     * @param executorService  An {@link ExecutorService} to run and control the computation and to perform the actions.
+     * @param executor  An {@link Executor} to run and control the computation and to perform the actions.
      * @param computation A (possibly blocking) computation
      * @param <T> value type of the Future
      * @return a new {@code FutureImpl} instance
      */
-    static <T> FutureImpl<T> async(ExecutorService executorService, CheckedConsumer<Predicate<Try<? extends T>>> computation) {
+    static <T> FutureImpl<T> async(Executor executor, CheckedConsumer<Predicate<Try<? extends T>>> computation) {
         // In a single-threaded context this Future may already have been completed during initialization.
-        return new FutureImpl<>(executorService, Option.none(), Queue.empty(), Queue.empty(), future -> executorService.submit(() -> {
+        return new FutureImpl<>(executor, Option.none(), Queue.empty(), Queue.empty(), tryComplete -> createThread(executor, () -> {
             try {
-                computation.accept(future::tryComplete);
+                computation.accept(tryComplete);
             } catch (Throwable x) {
-                future.tryComplete(Try.failure(x));
+                tryComplete.test(Try.failure(x));
             }
         }));
     }
 
+    private static Thread createThread(Executor executor, Runnable runnable) throws InterruptedException {
+        final AtomicReference<Thread> thread = new AtomicReference<>(null);
+        executor.execute(() -> {
+            synchronized (thread) {
+                thread.set(Thread.currentThread());
+                thread.notify();
```

Comment:
Using monitors in the presence of ForkJoinPool seems to deadlock the JVM (on the CI server). Instead of using a ManagedBlocker I will rewrite this without blocking.